### PR TITLE
 feat(ivy): support generating view and content queries

### PR DESCRIPTION
**Pull request by chuckjaz**
## PR Type
What kind of change does this PR introduce?

```
[x] Feature
```

## What is the current behavior?

View and content queries are not supported.

## What is the new behavior?

View and content queries are supported.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/compiler/src/output/output_ast.ts [hunk 1]
```diff
@@ -1004,7 +1004,7 @@ export class AstTransformer implements StatementVisitor, ExpressionVisitor {
   visitDeclareVarStmt(stmt: DeclareVarStmt, context: any): any {
     return this.transformStmt(
         new DeclareVarStmt(
-            stmt.name, stmt.value.visitExpression(this, context), stmt.type, stmt.modifiers,
+            stmt.name, stmt.value && stmt.value.visitExpression(this, context), stmt.type, stmt.modifiers,
```
Comment (vicb) [inline packages/compiler/src/output/output_ast.ts]: nit (personal pref): I like to extract things like `stmt.value && stmt.value.visitExpression(this, context)` or  `(this.value ? !!stmt.value && this.value.isEquivalent(stmt.value) : !stmt.value)` to a local const with a meaningful name.

File: packages/compiler/src/output/output_interpreter.ts [hunk 2]
```diff
@@ -95,7 +95,7 @@ class StatementInterpreter implements o.StatementVisitor, o.ExpressionVisitor {
   debugAst(ast: o.Expression|o.Statement|o.Type): string { return debugOutputAstAsTypeScript(ast); }
 
   visitDeclareVarStmt(stmt: o.DeclareVarStmt, ctx: _ExecutionContext): any {
-    ctx.vars.set(stmt.name, stmt.value.visitExpression(this, ctx));
+    ctx.vars.set(stmt.name, stmt.value && stmt.value.visitExpression(this, ctx));
```
Comment (vicb) [inline packages/compiler/src/output/output_interpreter.ts]: 1. is it ok to call `set` if no `stmt.value` ?
2. would it make sense not to call it ?

File: packages/compiler/src/output/ts_emitter.ts [hunk 3]
```diff
@@ -161,8 +161,10 @@ class _TsEmitterVisitor extends AbstractEmitterVisitor implements o.TypeVisitor
     }
     ctx.print(stmt, ` ${stmt.name}`);
     this._printColonType(stmt.type, ctx);
-    ctx.print(stmt, ` = `);
-    stmt.value.visitExpression(this, ctx);
+    if (stmt.value) {
+      ctx.print(stmt, ` = `);
```
Comment (vicb) [inline packages/compiler/src/output/ts_emitter.ts]: (super) nit: not consistent with JS (extra leading WS here)

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 4]
```diff
@@ -335,6 +342,33 @@ class TemplateDefinitionBuilder implements TemplateAstVisitor, LocalResolver {
       }
     }
 
+    // Define and update any view queries
+    if (this.viewQueries.length > 0) {
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: is the `if` needed here ?

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 5]
```diff
@@ -335,6 +342,33 @@ class TemplateDefinitionBuilder implements TemplateAstVisitor, LocalResolver {
...
+          args.push(this.outputCtx.importExpr(query.read.identifier!.reference));
+        }
+        this.instruction(this._creationMode, null, R3.query, ...args);
+
+        // (r3.qR(tmp = r3.ɵld(0)) && (ctx.someDir = tmp));
+        const temporary = this.temp();
+        const getQueryList = o.importExpr(R3.load).callFn([o.literal(querySlot)]);
+        const refresh = o.importExpr(R3.queryRefresh).callFn([temporary.set(getQueryList)]);
+        const updateDirective = o.variable(CONTEXT_NAME).prop(query.propertyName).set(
+          query.first ? temporary.key(o.literal(0)) : temporary);
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: Question: why 0 ?

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 6]
```diff
@@ -682,12 +737,73 @@ export function createFactory(
...
       o.INFERRED_TYPE, null, type.reference.name ? `${type.reference.name}_Factory` : null);
 }
 
+// Return a host binding function or null if one is not necessary.
+export function createHostBindingsFunction(type: CompileTypeMetadata, outputCtx: OutputContext,
+  queries: CompileQueryMetadata[]): o.Expression | null {
+
+  const statements: o.Statement[] = [];
+
+  const temporary = function () {
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: here I would have used `if (queries.length > 0) { temp = ...; ...}`

File: packages/compiler/src/render3/r3_view_compiler.ts [hunk 7]
```diff
@@ -682,12 +737,73 @@ export function createFactory(
...
+    const getQueryList = getDirectiveMemory.key(o.literal(index + 1));
+    const assignToTemporary = temporary().set(getQueryList);
+    const callQueryRefresh = o.importExpr(R3.queryRefresh).callFn([assignToTemporary]);
+    const updateDirective = getDirectiveMemory.key(o.literal(0, o.INFERRED_TYPE)).prop(query.propertyName).set(
+      query.first ? temporary().key(o.literal(0)) : temporary());
+    const andExpression = callQueryRefresh.and(updateDirective);
+    statements.push(andExpression.toStmt());
+  }
+
+  if (statements.length > 0) {
```
Comment (vicb) [inline packages/compiler/src/render3/r3_view_compiler.ts]: ... and you can remove also this if - if you apply the previous comment IIUC

File: packages/compiler/test/render3/r3_compiler_compliance_spec.ts [hunk 8]
```diff
@@ -284,6 +284,133 @@ describe('compiler compliance', () => {
...
+        const files = {
+          app: {
+          ...directive,
+          'view_query.component.ts': `
+            import {Component, NgModule, ViewChild} from '@angular/core';
+            import {SomeDirective} from './some.directive';
+
+            @Component({
+              selector: 'view-query-component',
+              template: \`
```
Comment (vicb) [inline packages/compiler/test/render3/r3_compiler_compliance_spec.ts]: nit: 1 line ?

File: packages/compiler/test/render3/r3_compiler_compliance_spec.ts [hunk 9]
```diff
@@ -284,6 +284,133 @@ describe('compiler compliance', () => {
...
+        const files = {
+          app: {
+          ...directive,
+          'spec.ts': `
+            import {Component, ContentChild, NgModule} from '@angular/core';
+            import {SomeDirective} from './some.directive';
+
+            @Component({
+              selector: 'content-query-component',
+              template: \`
```
Comment (vicb) [inline packages/compiler/test/render3/r3_compiler_compliance_spec.ts]: nit: 1 line

File: packages/compiler/test/render3/r3_compiler_compliance_spec.ts [hunk 10]
```diff
@@ -284,6 +284,133 @@ describe('compiler compliance', () => {
           result.source, ComplexComponentDefinition, 'Incorrect ComplexComponent definition');
     });
 
+    describe('queries', () => {
```
Comment (vicb) [inline packages/compiler/test/render3/r3_compiler_compliance_spec.ts]: missing a test for descendant ?

**Review by vicb — APPROVED** [review]
Approving as my comments are mostly questions/remarks. May be add a test ?

File: packages/compiler/test/render3/r3_compiler_compliance_spec.ts (hunk 10 above)
Comment (chuckjaz) [reply packages/compiler/test/render3/r3_compiler_compliance_spec.ts]: The test should be in the canonical specification. I believe Kara is adding it or should do that first.

File: packages/compiler/src/render3/r3_view_compiler.ts (hunk 5 above)
Comment (chuckjaz) [reply packages/compiler/src/render3/r3_view_compiler.ts]: This should be `prop('first')`

File: packages/compiler/src/output/ts_emitter.ts (hunk 3 above)
Comment (chuckjaz) [reply packages/compiler/src/output/ts_emitter.ts]: Fixed the JS version

File: packages/compiler/src/render3/r3_view_compiler.ts (hunk 6 above)
Comment (chuckjaz) [reply packages/compiler/src/render3/r3_view_compiler.ts]: As discussed offline, more code is coming in the a follow-on PR that will add more code here that will make the current structure make sense.

File: packages/compiler/src/render3/r3_view_compiler.ts (hunk 7 above)
Comment (chuckjaz) [reply packages/compiler/src/render3/r3_view_compiler.ts]: See above

File: packages/compiler/test/render3/r3_compiler_compliance_spec.ts (hunk 8 above)
Comment (chuckjaz) [reply packages/compiler/test/render3/r3_compiler_compliance_spec.ts]: Matches the canonical specification.

File: packages/compiler/test/render3/r3_compiler_compliance_spec.ts (hunk 9 above)
Comment (chuckjaz) [reply packages/compiler/test/render3/r3_compiler_compliance_spec.ts]: See above

**Comment by vicb** [comment]
http://test/OCL:186491140:BASE:186491883:1519240951446:7f71a10c

**Comment by vicb** [comment]
error: https://travis-ci.org/angular/angular/jobs/344109416#L1157

**Pull request closed without merging**